staging/src/k8s.io/apiserver/pkg/storage/storagecluster/tenantstoragemapmanager.go

```diff
@@ -0,0 +1,273 @@
+	ts.rev = rev
+	ts.mux.Unlock()
+	klog.V(4).Infof("mux released addTenant.")
+}
+
+func (ts *TenantStorageMapManager) updateTenant(old, cur interface{}) {
+	curTenant := cur.(*v1.Tenant)
+	oldTenant := old.(*v1.Tenant)
+
+	if curTenant.ResourceVersion == oldTenant.ResourceVersion {
+		return
+	}
+
+	oldRev, _ := strconv.ParseInt(oldTenant.ResourceVersion, 10, 64)
+	newRev, err := strconv.ParseInt(curTenant.ResourceVersion, 10, 64)
+	if err != nil {
+		klog.Errorf("Got invalid resource version %s for tenant %v.", curTenant.ResourceVersion, curTenant)
+		return
+	}
+
+	if diff.RevisionIsNewer(uint64(oldRev), uint64(newRev)) {
+		klog.V(2).Infof("Got staled tenant event %+v in UpdateFunc. Existing Version %s, new instance version %s.", curTenant, oldTenant.ResourceVersion, curTenant.ResourceVersion)
+		return
+	}
+
+	ts.mux.Lock()
+	klog.V(4).Infof("mux acquired updateTenant.")
+	defer func() {
+		ts.tenantToClusterMap[curTenant.Name] = &tenantStorage{
+			clusterId: curTenant.Spec.StorageClusterId,
+			rev:       newRev,
+		}
+		ts.rev = newRev
+		ts.mux.Unlock()
+		klog.V(4).Infof("mux released updateTenant.")
+	}()
+
+	if curTenant.Name != oldTenant.Name {
+		klog.Warningf("Tenant name was updated from %s to %s", oldTenant.Name, curTenant.Name)
+		delete(ts.tenantToClusterMap, oldTenant.Name)
+		return
+	}
+
+	existedTenant, isOK := ts.tenantToClusterMap[oldTenant.Name]
+	if !isOK {
+		klog.Warningf("Received event for update tenant %s not in tenant map", oldTenant.Name)
+		return
+	}
+	if curTenant.Spec.StorageClusterId != existedTenant.clusterId {
+		klog.Warningf("Received event for update tenant %s cluster. New cluster %s, old cluster %s. Currently storage moving is not supported.",
+			curTenant.Name, curTenant.Spec.StorageClusterId, existedTenant.clusterId)
+		return
+	}
+}
+
+func (ts *TenantStorageMapManager) deleteTenant(obj interface{}) {
+	tenant, ok := obj.(*v1.Tenant)
+	if !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("couldn't get object from tombstone %+v.", obj))
+			return
+		}
+		tenant, ok = tombstone.Obj.(*v1.Tenant)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("tombstone contained object that is not a tenant %#v.", obj))
+			return
+		}
+	}
+
+	klog.V(3).Infof("Received delete event for tenant [%+v]", tenant)
+	rev, err := strconv.ParseInt(tenant.ResourceVersion, 10, 64)
+	if err != nil {
+		klog.Errorf("Got invalid resource version %s for tenant %v.", tenant.ResourceVersion, tenant.Name)
+		return
+	}
+
+	if diff.RevisionIsNewer(uint64(ts.rev), uint64(rev)) {
+		klog.Errorf("Got invalid resource version %s for tenant %s.", tenant.ResourceVersion, tenant.Name)
+		return
+	}
+	ts.mux.Lock()
+	klog.V(4).Infof("mux acquired deleteTenant.")
+	delete(ts.tenantToClusterMap, tenant.Name)
+
+	ts.mux.Unlock()
+	klog.V(4).Infof("mux released deleteTenant.")
+}
+
+func (ts *TenantStorageMapManager) GetClusterIdFromTenant(tenant string) string {
+	ts.mux.RLock()
```

Comment:
It seems that read and write have no differences at all.

Reply:
https://golang.org/src/sync/rwmutex.go
The lock can be held by an arbitrary number of readers or a single writer.

RLock does not prevent additional reader, but it can prevent from writing.